src/persistence/sql.rs

```diff
@@ -1,37 +1,1008 @@
+
+                let current_amount = U256::from_str(&claimed_rewards).unwrap_or(U256::ZERO);
+                let new_claimed_rewards = current_amount
+                    .checked_add(U256::from(*amount))
+                    .ok_or_else(|| {
+                        anyhow::anyhow!(
+                            "Overflow: adding {amount} to claimed rewards {current_amount}"
+                        )
+                    })?;
+
+                // Update claimed rewards with the computed value
+                let result = sqlx::query(
+                    "UPDATE wallet
+                     SET claimed_rewards = $1
+                     WHERE address = $2",
+                )
+                .bind(new_claimed_rewards.to_string())
+                .bind(address.to_string())
+                .execute(&mut **tx)
+                .await?;
+
+                if result.rows_affected() != 1 {
+                    return Err(anyhow::anyhow!(
+                        "Expected to update 1 wallet row, but {} were affected",
+                        result.rows_affected()
+                    ))
+                    .map_err(Into::into);
+                }
+            }
+            WalletDiff::DelegatedToNode(delegation) => {
+                // Check if delegation exists
+                let existing = sqlx::query_as::<_, (String,)>(
+                    "SELECT amount FROM delegation WHERE delegator = $1 AND node = $2",
+                )
+                .bind(delegation.delegator.to_string())
+                .bind(delegation.node.to_string())
+                .fetch_optional(&mut **tx)
+                .await?;
+
+                let new_amount = if let Some((amount,)) = existing {
+                    let current_amount = U256::from_str(&amount).unwrap_or(U256::ZERO);
+                    current_amount.checked_add(delegation.amount)
+                        .ok_or_else(|| anyhow::anyhow!(
+                            "Overflow: adding delegation {} to existing amount {current_amount}",
+                            delegation.amount
+                        ))?
+                } else {
+                    delegation.amount
+                };
+
+                // upsert
+                sqlx::query(
+                    "INSERT INTO delegation (delegator, node, amount, unlocks_at, withdrawal_amount)
+                     VALUES ($1, $2, $3, 0, '0')
+                     ON CONFLICT(delegator, node) DO UPDATE SET
+                        amount = excluded.amount",
+                )
+                .bind(delegation.delegator.to_string())
+                .bind(delegation.node.to_string())
+                .bind(new_amount.to_string())
+                .execute(&mut **tx)
+                .await?;
+            }
+            WalletDiff::UndelegatedFromNode(withdrawal) => {
+                // Read current amount to calculate new amount after undelegation
+                let (amount,) = sqlx::query_as::<_, (String,)>(
```

Comment:
I think might as well select `withdrawal_amount` and `unlocks_at` here as well, and error if they are not both 0. The contract should prevent multiple pending withdrawals at the same time, but it seems like a good sanity check to have here